presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/DatabaseMetadataWriter.java

```diff
@@ -0,0 +1,718 @@
+                ShardWriterDao dao = handle.attach(ShardWriterDao.class);
+
+                // acquire shard lock
+                long abortedCommitId = getLockedAbortedCommitId(dao);
+
+                // skip if no rollback needed
+                if (abortedCommitId > commitId) {
+                    return;
+                }
+
+                dao.updateAbortedCommitId(commitId);
+
+                // rollback shard
+                dao.rollback(commitId);
+
+                for (long tableId : rollbackInfo.getWrittenTableIds()) {
+                    dao.rollbackCreatedIndexChunks(chunkIndexTable(tableId), commitId);
+                    dao.rollbackDeletedIndexChunks(chunkIndexTable(tableId), commitId);
+                }
+            });
+        }
+
+        // rollback indexes
+        for (long tableId : rollbackInfo.getCreatedTableIds()) {
+            shardDbi.get(getTableShard(tableId)).useHandle(handle ->
+                    dropIndexTable(handle, tableId));
+        }
+
+        for (TableColumn column : rollbackInfo.getAddedColumns()) {
+            shardDbi.get(getTableShard(column.getTableId())).useHandle(handle ->
+                    dropIndexTableColumn(databaseType, handle, column.getTableId(), column.getColumnId()));
+        }
+
+        // finish rollback
+        masterDbi.useTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            getLockedCurrentCommitId(dao);
+
+            // verify we are still rolling back
+            ActiveCommit commit = dao.getActiveCommit();
+            if ((commit == null) || (commit.getCommitId() != commitId)) {
+                return;
+            }
+            verifyMetadata(commit.isRollingBack(), "Commit should be rolling back");
+
+            // finalize rollback
+            deleteActiveCommit(dao);
+        });
+    }
+
+    @Override
+    public long beginCommit()
+    {
+        return masterDbi.inTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            long currentCommitId = getLockedCurrentCommitId(dao);
+
+            // verify no commit in progress
+            if (dao.getActiveCommit() != null) {
+                throw new PrestoException(RAPTOR_INTERNAL_ERROR, "Commit already in progress");
+            }
+
+            // allocate commit ID
+            long commitId = sequenceManager.nextValue("commit_id", 1);
+            verifyMetadata(commitId > currentCommitId, "Invalid next commit ID: %s", commitId);
+
+            // start commit
+            dao.insertActiveCommit(commitId, System.currentTimeMillis());
+
+            return commitId;
+        });
+    }
+
+    @Override
+    public void finishCommit(long commitId, OptionalLong transactionId)
+    {
+        writeMaster(commitId, (dao, commit) -> {
+            deleteActiveCommit(dao);
+
```

Comment:
If JVM crashes here,  recover won’t do anything, since recover seems only based on active_commit content.
But these 3 are not executed:
insertCommit
updateCurrentCommit
finalizeTransaction
So they are in some wrong state. 1st question is do we worry about it? It looks to me the 3 things are not used. Maybe it’s fine. For ex, table “transactions”, column “successful”. Only written, but I didn't find a place where it's read.

If it’s fine, another question is shall we get rid of these 3 things. Get rid of column “successful”, since it’s not used.